### DISCUSSION TITLE
[RFC] 033 - 模型列表

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#1686] [RFCs]
## 背景

1. OpenRouter 具有大量的模型 #1684
2. Ollama 也支持大量的模型
3. 纯文本语法的模型自定义方案体验欠佳 #1157
4. 目前每个 Provider 中的 ModelCard 都预先定义了很多模型，但是默认隐藏了，用户需要启用的话得自行配置自定义模型字符串，不太方便
5. 每个 provider 都会有新模型出现，我们的跟进速度不一定及时，需要允许用户添加自定义模型
6. 用户会希望设定自定义模型能力，例如 #1517, #1244 ，除了环境变量方案，最好我们也能提供界面配置的能力。

## 设计思路

实现一个统一的模型选择器的设计方案，同时满足以下需求：
1. 支持从模型列表中选择模型进行展示，不用通过字符串输入(解决 3 和 4 的问题)
2. 模型列表支持预定义和通过请求获取，默认采用预定义，请求获取为增强项 (解决 1 和 2)
3. 支持用户通过输入字符串来创建自定义模型，并且自行定义模型能力 (解 5 和 6)

#### UI

使用 antd 的 [Select tag](https://ant-design.antgroup.com/components/select-cn#select-demo-tags) 模式，可以同时兼容 1 和 3 的需求。

![image](https://github.com/lobehub/lobe-chat/assets/28616219/0fb84f69-39d3-4bbb-8f9e-aa1c9dea5864)

#### 数据流

模型列表的展示从颗粒度上应该分为三个层次：
1. 每一个 Provider 的完整 `ProviderModelList` ，这个列表的数据来源可能是LobeChat 预定义，也可能是 OpenRouter/Ollama 这样有一个远程数据；
2. 用户在 settings 中 enable 的 `models` ，这个数据代表着用户希望在界面中开启的模型，比如某些用户就不希望显示 `claude2.1` （[refs](https://github.com/lobehub/lobe-chat/issues/1858)），此处可以只记录 id ；
3. 会话界面的模型选择 `model` ，这个字段代表在会话中实际使用的模型。

之前的数据处理由于不存在 `models` 这一层概念，因此 model 展示部分是直接在 `modelSelectList` 中完成全部处理工作：
1. 服务端 customModelName 、用户侧 customModelName 整合；
2. customModelName 到 modelCard 转换
3. 展示过滤

由于现在在 Settings 中多了一层 models 的控制，我们可以尝试优化一下这部分的实现了：

1. 服务端 `customModelName` 应该仍然可以保留，仍然作为服务端控制 ProviderModelList 的方法；
2. 客户端的 `customModelName` 可以直接移除，替换为 `models` 的使用逻辑，这里就需要做一个数据迁移，并看下对齐到  customModelName 还需要的特性；

`modelProvider.providerModelList` 代表了 lobechat + server 预定义的官方 list。

`modelSelectList` 代表了经过加工后的，用户偏好的模型选择列表。

加工所需要的第一个字段：

- `models`: 用于处理用户开启/关闭某个选择的模型

加工所需要的第二个字段：

- `customModelCards`: 用于配置模型层的模型列表， 作用就是在 client 侧补充一个定义模型属性/能力的位置

## 进展

在 https://github.com/lobehub/lobe-chat/pull/1916 实施

**Comment by wuyuan1992** [comment]
各提供商包括 Ollama， 都提供了模型列表查询的接口。
- 能否考虑通过定时更新和进入语言模型设置界面时查询并更新提供商接口？
- 再在这个基础上可以允许用户设置，比如：
1. 设置别名
2. 隐藏指定模型
3. 当模型标识为多模态时，可以手动设置是否支持 vision
4. 是否支持函数调用等
